libmultipath/public_api.c

```diff
@@ -0,0 +1,794 @@
+{
+	struct dmmp_mpath **mpaths = NULL;
+	uint32_t mpath_count = 0;
+	uint32_t i = 0;
+	struct dmmp_mpath *rc_mpath = NULL;
+
+	if (mpath_name == NULL)
+		return NULL;
+
+	if (dmmp_mpath_list(&mpaths, &mpath_count) != 0)
+		goto out;
+
+	for (i = 0; i < mpath_count; ++i) {
+		if (mpaths[i]->alias == NULL)
+			continue;
+		if (strcmp(mpath_name, mpaths[i]->alias) == 0) {
+			rc_mpath = mpaths[i];
+			goto out;
+		}
+	}
+
+out:
+	if (rc_mpath != NULL)
+		rc_mpath = dmmp_mpath_copy(rc_mpath);
+
+	dmmp_mpath_list_free(mpaths, mpath_count);
+
+	return rc_mpath;
+}
+
+struct dmmp_mpath *dmmp_mpath_copy(struct dmmp_mpath *mpath)
+{
+	struct dmmp_mpath *mpath_new = NULL;
+	struct dmmp_path_group *mp_pg_new = NULL;
+	struct dmmp_path_group *mp_pg_old = NULL;
+	struct dmmp_path *mp_path_new = NULL;
+	struct dmmp_path *mp_path_old = NULL;
+	uint32_t i = 0;
+	uint32_t j = 0;
+
+	if (mpath == NULL)
+		return NULL;
+
+	mpath_new = _dmmp_mpath_new();
+	if (mpath_new == NULL)
+		return NULL;
+
+	memcpy(mpath_new, mpath, sizeof(struct dmmp_mpath));
+	mpath_new->mp_pgs = NULL;
+	mpath_new->alias = NULL;
+	mpath_new->mp_pg_count = 0;
+
+	if (mpath->alias != NULL) {
+		mpath_new->alias = STRDUP(mpath->alias);
+		if (mpath_new->alias == NULL)
+			goto error_out;
+	}
+
+	if (mpath->mp_pg_count == 0)
+		return mpath_new;
+
+	mpath_new->mp_pgs = (struct dmmp_path_group **)
+		MALLOC(sizeof(struct dmmp_path_group *) * mpath->mp_pg_count);
+
+	if (mpath_new->mp_pgs == NULL)
+		goto error_out;
+
+	mpath_new->mp_pg_count = mpath->mp_pg_count;
+
+	for (i = 0; i < mpath->mp_pg_count; ++i) {
+		mp_pg_old = mpath->mp_pgs[i];
+		mp_pg_new = _dmmp_path_group_new();
+		if (mp_pg_new == NULL)
+			goto error_out;
+		mpath_new->mp_pgs[i] = mp_pg_new;
+
+		memcpy(mp_pg_new, mp_pg_old, sizeof(struct dmmp_path_group));
+		mp_pg_new->selector = NULL;
+		mp_pg_new->mp_paths = NULL;
+		mp_pg_new->mp_path_count = 0;
+
+		mp_pg_new->selector = STRDUP(mp_pg_old->selector);
+		if (mp_pg_new->selector == NULL)
+			goto error_out;
+		if (mp_pg_old->mp_path_count == 0)
+			continue;
+
+		mp_pg_new->mp_paths = (struct dmmp_path **)
+			MALLOC(sizeof(struct dmmp_path *) *
+			       mp_pg_old->mp_path_count);
+		if (mp_pg_new->mp_paths == NULL)
+			goto error_out;
+
+		mp_pg_new->mp_path_count = mp_pg_old->mp_path_count;
+
+		for (j = 0; j < mp_pg_old->mp_path_count; ++j) {
+			mp_path_old = mp_pg_old->mp_paths[j];
+			mp_path_new = _dmmp_path_new();
+			if (mp_path_new == NULL)
+				goto error_out;
+			mp_pg_new->mp_paths[j] = mp_path_new;
+
+			memcpy(mp_path_new, mp_path_old,
+			       sizeof(struct dmmp_path));
+		}
+	}
+
+	return mpath_new;
+
+error_out:
+	dmmp_mpath_free(mpath_new);
+	return NULL;
+}
+
+int dmmp_path_group_list_get(struct dmmp_mpath *mpath,
+			     struct dmmp_path_group ***mp_pgs,
+			     uint32_t *mp_pg_count)
+{
+	*mp_pgs = NULL;
+	*mp_pg_count = 0;
+	if (mpath != NULL)
+		*mp_pgs = mpath->mp_pgs;
+		*mp_pg_count = mpath->mp_pg_count;
+	return 0;
```

Comment:
Maybe you are planning on returning an error in the case mp_pgs || mp_pg_count are NULL?  If not change to return void?